R/mobr.R

```diff
@@ -10,221 +10,244 @@ require(pracma)
-    ScaleBy = NA) {
+null_N = function(dat_sp, dat_plot, groups, nperm = 1000, CI = 0.95, ScaleBy = NA) {
     # This function generates null confidence intervals for the deltaSN curve
     # assuming that the two treatments do not differe systematically in N.
-    avg_dens = get_avg_dens(dat_sp, dat_plot, ScaleBy)
-    nplots = c(nrow(dat_plot[dat_plot[ , 2] == treatment1, ]), 
-               nrow(dat_plot[dat_plot[ , 2] == treatment2, ]))
-    dat_plot_trmts = dat_plot[dat_plot[ , 2] %in% c(treatment1, treatment2), ]
-    # Only keep the plots for the two given treatments
-    dat_sp = dat_sp[match(dat_plot_trmts[ , 1], dat_sp[ , 1]), ]  
-    trmt_sads = list()
-    for (trmt in c(treatment1, treatment2)) {
-        plots = dat_plot[dat_plot[ , 2] == trmt, 1]
-        trmt_sad = apply(dat_sp[dat_sp[ , 1] %in% plots, 2:ncol(dat_sp)], 2, sum)
-        trmt_sads = c(trmt_sads, list(rep(2:ncol(dat_sp), trmt_sad)))
+    S = ncol(dat_sp)
+    dat_sp = dat_sp[dat_plot$group %in% groups, ]
+    dat_plot = dat_plot[dat_plot$group %in% groups, ]
+    Nplot = min(table(dat_plot$group))
+    grp_sads = list()
+    for (i in groups) {
+        plots = dat_plot$plot[dat_plot$group == i]
+        grp_sad = colSums(dat_sp[row.names(dat_sp) %in% plots, ])
+        grp_sads[[i]] = rep(1:S, grp_sad)
     }
     # Abundance within each plot
-    n_plot = apply(dat_sp[ , 2:ncol(dat_sp)], 1, sum)  
-    sad_row = lapply(1:nrow(dat_sp), function(x) 
-                     rep(2:ncol(dat_sp), dat_sp[x, 2:ncol(dat_sp)]))
-    Nind = min(floor(min(nplots) * avg_dens), 
-               length(unlist(trmt_sads[1])), 
-               length(unlist(trmt_sads[2])))
+    plot_abus = apply(dat_sp, 1, sum)  
+    Nind = min(sapply(grp_sads, length)) 
     deltaSN_perm = matrix(NA, nperm, Nind)
     for (i in 1:nperm) {
-        n_plot_shuffle = sample(n_plot)
-        dat_sp_perm = as.data.frame(matrix(0, nrow(dat_sp), ncol(dat_sp)))
-        dat_sp_perm[ , 1] = dat_sp[ , 1]
-        new_counts = sapply(1:nrow(dat_sp), function(x) as.numeric(table(factor(
-                            sample(if (length(unlist(sad_row[x])) > 0) 
-                                       unlist(sad_row[x])  # If plot is empty, use treatment-level SAD
-                                   else
-                                       unlist(trmt_sads[
-                                           which(c(treatment1, treatment2) == 
-                                                 dat_plot[x, 2])]),
-                                   n_plot_shuffle[x], replace = T),
-                            levels = 2:ncol(dat_sp)))))
-        dat_sp_perm[ , 2:ncol(dat_sp_perm)] = as.data.frame(t(new_counts))
-        deltaSN_perm[i, ] = get_deltaSN(dat_sp_perm, dat_plot_trmts,
-                                        treatment1, treatment2, ScaleBy)[1:Nind]
+        plot_abu_perm = sample(plot_abus)
+        sp_draws = sapply(1:nrow(dat_sp), function(x) 
```

Comment:
We probably want to include the case where dat_sp[x, ] are all zeros (see old code).